src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs

```diff
@@ -450,7 +450,13 @@ private unsafe void CloseRequestQueueHandle()
             if ((_requestQueueHandle != null) && (!_requestQueueHandle.IsInvalid))
             {
                 if (NetEventSource.IsEnabled) NetEventSource.Info($"Dispose ThreadPoolBoundHandle: {_requestQueueBoundHandle}");
-                _requestQueueBoundHandle?.Dispose();
+
+                if (_requestQueueBoundHandle != null)
+                {
+                    _requestQueueBoundHandle.Dispose();
+                    _requestQueueBoundHandle = null;
```

Comment:
It looks like setting this to null could, with certain interleavings, cause:
https://github.com/dotnet/corefx/blob/7a668f4f13b962a20a4585fffadbf537d5ebf570/src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs#L165
to return null.  Might this cause a null reference exceptions elsewhere?  Callers of that property expect it to always return non-null.

Reply:
I think you're right; if someone calls `Stop()` at just the right moment, `ThreadPoolBoundHandle` might return null.

I can fix this -- if I include the `return _requestQueueBoundHandle` in the lock, then we won't NRE due to this. It creates a new situation where we might allocate a new one that doesn't get disposed, but GC will take care of that and Dispose is a no-op right now so that is fine regardless.

This is related to one of those race conditions I alluded to. With this fixed, those other things will still get an error because they all try to, immediately after getting the `ThreadPoolBoundHandle`, use the `_requestQueueHandle` that is being disposed right after the `_requestQueueBoundHandle`. (╯°□°）╯︵ ┻━┻

Reply:
>(╯°□°）╯︵ ┻━┻

What is that? :)

Reply:
https://knowyourmeme.com/memes/flipping-tables-╯□╯︵-┻━┻?full=1

Reply:
Thanks.  Obviously I need to take an updated emoticon course to learn these things. :)

Reply:
> This is related to one of those race conditions I alluded to. With this fixed, those other things will still get an error because they all try to, immediately after getting the ThreadPoolBoundHandle, use the _requestQueueHandle that is being disposed right after the _requestQueueBoundHandle.

So does that mean this PR won't make things worse (and obviously will make one case better), or are there cases where this will turn things that were previously nops into failures?

Reply:
> So does that mean this PR won't make things worse (and obviously will make one case better), or are there cases where this will turn things that were previously nops into failures?

The basic race when stopping in HttpListener is:
* An outstanding multi-step op completes its first leg successfully.
* You call Stop() and close everything.
* The outstanding op starts its second leg and hopefully the handle was set to `NULL` or `INVALID_HANDLE` so the http.sys function can error out.

This is what exists today.

In .NET 4.8 there's an additional race, which doesn't effect .NET Core because we don't support restarting:

* An outstanding multi-step op completes its first leg successfully and right before its second leg, the thread ends its quantum.
* You call Stop() and close everything.
* You call Start() again.
* Thread resumes and the outstanding op starts its second leg and it passes now invalid state to http.sys functions.

This PR will re-enable this 2nd race.